skbio/alignment/_tabular_msa.py

```diff
@@ -717,6 +707,92 @@ def reindex(self, key=None, keys=None):
+            If None, the MSA is checked to see if a previous minter has been
+            cached. If not, an exception is raised. If a cached key is found
+            it is used to reindex the MSA.
+
+        Raises
+        ------
+        TypeError
+            If the sequence object is a type that doesn't have an alphabet
+        TypeError
+            If the type of the sequence does not match the dtype of the MSA.
+        ValueError
+            If the length of the sequence does not match the number of
+            positions in the MSA.
+        OperationError
+            If no key is provided, but the MSA has keys, and no cached key
+            is found.
+
+        See Also
+        --------
+        reindex
+
+        Notes
+        -----
+        The MSA is not automatically re-aligned when a sequence is appended.
+        Therefore, this operation is not necessarily meaningful on its own.
+
+        Examples
+        --------
+        >>> from skbio import DNA, TabularMSA
+        >>> msa = TabularMSA([DNA('')])
+        >>> msa.append(DNA(''))
+        >>> msa == TabularMSA([DNA(''), DNA('')])
+        True
+
+        >>> msa = TabularMSA([DNA('', metadata={'id': 'a'})], minter='id')
+        >>> msa.append(DNA('', metadata={'id': 'b'}), minter='id')
+        >>> msa == TabularMSA([DNA('', metadata={'id': 'a'}),
+        ...                    DNA('', metadata={'id': 'b'})], minter='id')
+        True
+        """
+        if minter is None:
+            if self.has_keys():
+                minter = self.minter
+        else:
+            if not self.has_keys():
+                raise OperationError(
+                    "key was provided but MSA does not have keys.")
+
+        self._add_sequence(sequence)
+        self.reindex(minter=minter)
+
+    def _add_sequence(self, sequence):
+        msa_is_empty = not len(self)
+        if msa_is_empty:
+            dtype = type(sequence)
+            if not issubclass(dtype, IUPACSequence):
+                raise TypeError(
+                    "`sequence` must be a scikit-bio sequence object "
+                    "that has an alphabet, not type %r" % dtype.__name__)
+            self._dtype = dtype
+            self._shape = _Shape(sequence=1, position=len(sequence))
+            self._seqs = [sequence]
+        elif type(sequence) is not self.dtype:
+            raise TypeError(
+                "`sequence` Must match the type of any other sequences "
+                "already in the MSA. Type %r does not match type %r" %
+                (type(sequence).__name__, self.dtype.__name__))
+        elif len(sequence) != self.shape.position:
+            raise ValueError(
+                "`sequence` length must match the length of any other "
+                "sequences already in the MSA: %r != %r"
+                % (len(sequence), self.shape.position))
+        else:
+            self._shape = _Shape(sequence=self._shape.sequence + 1,
```

Comment:
Setting the shape is happening in three places. Can you modify the `shape` property to construct a `Shape` object based on the number of rows and positions in the first sequence? That way this book-keeping won't be necessary.